File: algos/lowest-common-ancestor-bst/lowest-common-ancestor-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function (root, p, q) {
  // note - this solution only works if the p and q nodes are guaranteed to be present in the tree

  // both p and q are in the left subtree
  if (root.val > p.val && root.val > q.val) {
    return lowestCommonAncestor(root.left, p, q);
  }
  // both p and q are in the right subtree
  if (root.val < p.val && root.val < q.val) {
    return lowestCommonAncestor(root.right, p, q);
  }

  // if the p and q nodes are not guaranteed to be in the tree, here we will confirm with a dfs.
  return root;
};

File: algos/lowest-common-ancestor-bst/lowest-common-ancestor-bst.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	// both nodes are in left subtree
	if root.Val > p.Val && root.Val > q.Val {
		return lowestCommonAncestor(root.Left, p, q)
	}
	// both nodes are in right subtree
	if root.Val < p.Val && root.Val < q.Val {
		return lowestCommonAncestor(root.Right, p, q)
	}
	// assume p node in left sub and q node in right sub
	return root
}